File: VoCal/VoCal/Persistence.swift
//
//  Persistence.swift
//  VoCal
//
//  JSON snapshot of the full app state to the Documents directory. Saved on
//  every mutation; loaded once at launch. Force-quit / cold-launch resumes
//  exactly where the user left off — no more "your day is gone" footgun.
//
//  We deliberately use plain Codable + a single file rather than SwiftData
//  because:
//   (1) zero Xcode-side ceremony (no @Model, no ModelContainer setup),
//   (2) easy to inspect via the file system during demos and bug reports,
//   (3) lets `AppModel` stay the single source of truth without async hops.
//
//  Schema is versioned. When we change shape we bump `version` and either
//  migrate or discard old payloads.
//

import Foundation

struct AppStateSnapshot: Codable {
    var version: Int
    var totals: DailyTotals
    var meals: [MealEntry]
    var profile: UserProfile
    var bodyMetrics: [BodyMetric]
    var coachMessages: [CoachMessage]
    var hasCompletedOnboarding: Bool
    var savedAt: Date

    static let currentVersion = 1
}

enum Persistence {
    private static let fileName = "vocal_state.v1.json"

    /// Returns the URL inside the app's Documents directory. Documents persists
    /// across launches and is included in iCloud Drive backups by default.
    private static func fileURL() -> URL? {
        guard let dir = try? FileManager.default.url(
            for: .documentDirectory, in: .userDomainMask,
            appropriateFor: nil, create: true
        ) else { return nil }
        return dir.appendingPathComponent(fileName)
    }

    /// Writes a snapshot. Best-effort — failures are logged via `print` and
    /// otherwise swallowed; we never want a save error to break a save flow.
    static func save(_ snapshot: AppStateSnapshot) {
        guard let url = fileURL() else { return }
        do {
            let encoder = JSONEncoder()
            encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
            encoder.dateEncodingStrategy = .iso8601
            let data = try encoder.encode(snapshot)
            // Atomic write so a crash mid-save doesn't leave a half-written file.
            try data.write(to: url, options: [.atomic, .completeFileProtection])
        } catch {
            print("[VoCal.Persistence] save failed: \(error)")
        }
    }

    /// Loads the most recent snapshot or nil if no save exists / decode fails.
    /// On schema-version mismatch we return nil and let the caller fall back
    /// to a fresh empty state (rather than risk a corrupt-but-loadable mix).
    static func load() -> AppStateSnapshot? {
        guard let url = fileURL(),
              FileManager.default.fileExists(atPath: url.path),
              let data = try? Data(contentsOf: url) else { return nil }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        guard let snap = try? decoder.decode(AppStateSnapshot.self, from: data) else { return nil }
        if snap.version != AppStateSnapshot.currentVersion { return nil }
        return snap
    }

    /// Removes the persisted state. Useful for "Sign out" / reset flows.
    /// No-op if no file exists.
    static func clear() {
        guard let url = fileURL() else { return }
        try? FileManager.default.removeItem(at: url)
    }
}

extension AppStateSnapshot {
    /// Empty state for a brand-new user. Sensible defaults that survive
    /// onboarding immediately if the user skips fields.
    static func empty() -> AppStateSnapshot {
        AppStateSnapshot(
            version: AppStateSnapshot.currentVersion,
            totals: DailyTotals(
                date: .now,
                calorieGoal: 2000,
                caloriesEaten: 0,
                proteinGoal: 140,
                proteinEaten: 0,
                carbsGoal: 220,
                carbsEaten: 0,
                fatGoal: 65,
                fatEaten: 0
            ),
            meals: [],
            profile: UserProfile(
                displayName: "",
                streakDays: 0,
                weightLbs: 0,
                heightInches: 0,
                dailyCalorieGoal: 2000,
                sex: "",
                birthYear: 1995,
                entitlement: .free
            ),
            bodyMetrics: [],
            coachMessages: [],
            hasCompletedOnboarding: false,
            savedAt: .now
        )
    }
}

File: vocal_flutter/lib/services/persistence.dart
// JSON snapshot of the full app state to the app documents directory.
// Saved on every mutation; loaded once at launch. Mirrors Persistence.swift.

import 'dart:convert';
import 'dart:io';

import 'package:path_provider/path_provider.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../models/models.dart';

class Persistence {
  static const _fileName = 'vocal_state.v1.json';

  static Future<File?> _file() async {
    try {
      final dir = await getApplicationDocumentsDirectory();
      return File('${dir.path}/$_fileName');
    } catch (_) {
      return null;
    }
  }

  static Future<void> save(AppStateSnapshot snapshot) async {
    try {
      final f = await _file();
      if (f == null) return;
      // Write to a temp file then rename — atomic via the POSIX rename, so a
      // crash between write and rename leaves the previous good file intact.
      // Mirrors iOS Persistence.save's `.atomic` option.
      final tmp = File('${f.path}.tmp');
      await tmp.writeAsString(snapshot.encode(), flush: true);
      await tmp.rename(f.path);
    } catch (e) {
      // Best-effort: never let a save error break a save flow.
      // ignore: avoid_print
      print('[VoCal.Persistence] save failed: $e');
    }
  }

  static Future<AppStateSnapshot?> load() async {
    try {
      final f = await _file();
      if (f == null) return null;
      if (await f.exists()) {
        final raw = await f.readAsString();
        return AppStateSnapshot.decode(raw);
      }
      // Crash-recovery: a previous save may have been killed between
      // writeAsString and rename, leaving a .tmp orphan. If the main file is
      // missing but a complete tmp exists, prefer that over treating the user
      // as a fresh install (which would zero their day).
      final tmp = File('${f.path}.tmp');
      if (await tmp.exists()) {
        final raw = await tmp.readAsString();
        final decoded = AppStateSnapshot.decode(raw);
        if (decoded != null) {
          // Promote tmp → main so the next save uses the normal path.
          try { await tmp.rename(f.path); } catch (_) {}
          return decoded;
        }
      }
      return null;
    } catch (_) {
      return null;
    }
  }

  static Future<void> clear() async {
    try {
      final f = await _file();
      if (f != null && await f.exists()) await f.delete();
    } catch (_) {}
  }
}

/// Compact today snapshot mirrored to shared_preferences (UserDefaults
/// equivalent). Read by anything that wants today's macros without spinning
/// up the full UI. Includes the stale-day guard from DailyMacrosSnapshot.
class DailyMacrosSnapshot {
  static const _key = 'vocal.dailyMacrosSnapshot.v1';

  static Future<void> writeFrom(DailyTotals t) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(
        _key,
        jsonEncode({
          'date': t.date.toIso8601String(),
          'calorieGoal': t.calorieGoal,
          'caloriesEaten': t.caloriesEaten,
          'proteinGoal': t.proteinGoal,
          'proteinEaten': t.proteinEaten,
          'carbsGoal': t.carbsGoal,
          'carbsEaten': t.carbsEaten,
          'fatGoal': t.fatGoal,
          'fatEaten': t.fatEaten,
        }),
      );
    } catch (_) {}
  }

  static Future<Map<String, dynamic>?> read() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final raw = prefs.getString(_key);
      if (raw == null) return null;
      final j = jsonDecode(raw) as Map<String, dynamic>;
      final date = DateTime.tryParse(j['date'] as String? ?? '');
      final now = DateTime.now();
      final isToday = date != null &&
          date.year == now.year &&
          date.month == now.month &&
          date.day == now.day;
      if (!isToday) {
        return {
          'date': now.toIso8601String(),
          'calorieGoal': j['calorieGoal'],
          'caloriesEaten': 0,
          'proteinGoal': j['proteinGoal'],
          'proteinEaten': 0,
          'carbsGoal': j['carbsGoal'],
          'carbsEaten': 0,
          'fatGoal': j['fatGoal'],
          'fatEaten': 0,
        };
      }
      return j;
    } catch (_) {
      return null;
    }
  }
}
